### PR TITLE
[CALCITE-5634] Enable GREATEST, LEAST functions in PostgreSQL library

### PR DESCRIPTION
**Pull request by normanj-bitquill**
* Add new GREATEST and LEAST functions for PostgreSQL
* The new functions only return null if all arguments are null
* Created a new library for RedShift since it is not truly a union of Oracle and PostgreSQL functions

### CODE REVIEW

**Comment by mihaibudiu** [comment]
What's the relationship with #3142?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java [hunk 1]
```diff
@@ -72,6 +72,8 @@ public enum SqlLibrary {
   /** A collection of operators that are in PostgreSQL but not in standard
    * SQL. */
   POSTGRESQL("p", "postgresql"),
+  /** A collection of operators that are in RedShift but not in standard SQL. */
+  REDSHIFT("r", "redshift"),
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java:76]: This seems only tangentially related to this PR.
I wonder whether this deserves a separate issue and PR.

**Review by mihaibudiu — CHANGES_REQUESTED** [review]
I haven't read everything, but I think you should break this PR into (at least) two separate PRs:
- adding support for Redshift
- implementing GREATEST, LEAST
You probably need to create a new issue for redshift.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java (hunk 1 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java:76]: @mihaibudiu I ran into a problem with this:
https://github.com/apache/calcite/blob/main/babel/src/test/java/org/apache/calcite/test/BabelQuidemTest.java#L110

Both Oracle and PostgreSQL have `GREATEST` and `LEAST` functions but they behave differently. I needed some way for RedShift to be able to choose only one implementation.

**Comment by normanj-bitquill** [comment]
@mihaibudiu I'll split this as you suggest. The RedShift library change will go into a new PR.

**Comment by normanj-bitquill** [comment]
@mihaibudiu This PR is now based on top of
https://github.com/apache/calcite/pull/3829

When 3829 is merged in, I can rebase this off of main.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java (hunk 1 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibrary.java:76]: This has been moved to another PR.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 2]
```diff
@@ -101,7 +102,7 @@ private SqlLibraryOperators() {
    * converts the timezone of {@code datetime} from {@code tz1} to {@code tz2}.
    * This function is only on Redshift, but we list it in PostgreSQL
    * because Redshift does not have its own library. */
-  @LibraryOperator(libraries = {POSTGRESQL})
+  @LibraryOperator(libraries = {POSTGRESQL, REDSHIFT})
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:105]: I assume all these REDSHIFT should also move to the other PR.

File: site/_docs/reference.md [hunk 3]
```diff
@@ -2707,18 +2708,18 @@ In the following:
...
 | m p | CONCAT_WS(separator, str1 [, string ]*)      | Concatenates one or more strings, returns null only when separator is null, otherwise treats null arguments as empty strings
 | q | CONCAT_WS(separator, str1, str2 [, string ]*)  | Concatenates two or more strings, requires at least 3 arguments (up to 254), treats null arguments as empty strings
 | m | COMPRESS(string)                               | Compresses a string using zlib compression and returns the result as a binary string
 | b | CONTAINS_SUBSTR(expression, string [ , json_scope =&gt; json_scope_value ]) | Returns whether *string* exists as a substring in *expression*. Optional *json_scope* argument specifies what scope to search if *expression* is in JSON format. Returns NULL if a NULL exists in *expression* that does not result in a match
 | q | CONVERT(type, expression [ , style ])          | Equivalent to `CAST(expression AS type)`; ignores the *style* operand
-| p | CONVERT_TIMEZONE(tz1, tz2, datetime)           | Converts the timezone of *datetime* from *tz1* to *tz2*
+| p r | CONVERT_TIMEZONE(tz1, tz2, datetime)         | Converts the timezone of *datetime* from *tz1* to *tz2*
```
Comment (mihaibudiu) [inline site/_docs/reference.md:2722]: I also assume all these "r" changes will move to the other PR as well.

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java [hunk 4]
```diff
@@ -1900,6 +1902,60 @@ private static class GreatestConvertlet implements SqlRexConvertlet {
     }
   }
 
+  /** Convertlet that converts {@code GREATEST} and {@code LEAST}. */
+  private static class GreatestPgConvertlet implements SqlRexConvertlet {
+    @Override public RexNode convertCall(SqlRexContext cx, SqlCall call) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: this expansion is quadratic in the list size.
is this something we should worry about?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:105]: They will for everything except `GREATEST` and `LEAST`, since Redshift should also use the new implementation.

File: site/_docs/reference.md (hunk 3 above)
Comment (normanj-bitquill) [reply site/_docs/reference.md:2722]: That is correct. For this PR, they will only be added to `GREATEST` and `LEAST`.

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: I based this on the existing `GreatestConvertlet` class. It converts in a very similar way. The original `GreatestConvertlet` was written 8 years ago, so a different solution may make more sense now.

In general, the user calls the function with a list of scalars, such as `GREATEST(1, 2, 3, 4)`. There must be an upper limit, since it must fit in a SQL call that can be parsed.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: There are some PRs attempting to solve https://issues.apache.org/jira/projects/CALCITE/issues/CALCITE-3094, which happens when a table has too many columns. So what we assume as a small constant can actually be big.

I could see cases where compiler-generated code could produce a long list as an argument for GREATEST. 
So I wonder whether throwing an exception for a very long list of arguments (e.g., >100) would make sense.

**Comment by mihaibudiu** [comment]
In any case, I will review this again after it's rebased on #3829

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: It would be preferable to give a useful error message to the caller, such as "too many arguments". I don't have a good idea what would make sense for the limit.

I tested `CONCAT` in PostgreSQL and it gives an error after 100 arguments. It probably makes sense to bring that limitation into Calcite.

I also tested `GREATEST` in PostgreSQL. PostgreSQL must do something clever with the arguments, since it works up to at least 1001 arguments.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: The problem is that you want to compile to sql. In you use a language with local variables you can do it in size O(N). And postgres is not constrained to compile to sql.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 2 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: This has been cleaned up. No need to specify Redshift here since it will by default get PostgreSQL functions.

File: site/_docs/reference.md (hunk 3 above)
Comment (normanj-bitquill) [reply site/_docs/reference.md]: This is now cleaned up and only GREATEST and LEAST have changes.

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: An expression like `GREATEST(1, 2, ...)` is compiled and becomes a `SqlCall` object with arguments. The convertlet will then replace the `SqlCall` object with a new `SqlCall` for `CASE`. We are only manipulating objects after they are parsed, we do not generate SQL text and parse it again.

This will still explode original expression and use potentially much more memory.

With simple integer arguments, this PR works with at least 1000 values.

If no limit is in place, the user is still restricted by the max query length and what will fit in memory after the convertlet is run.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: yes, your intermediate language is as expressive as SQL. That's what I mean.
But if you have let expressions you can do much better.

**Review by mihaibudiu — APPROVED** [review]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 5]
```diff
@@ -10747,8 +10747,30 @@ void assertSubFunReturns(boolean binary, String s, int start,
...
+      f.checkString("greatest('on', 'earth')", "on   ", "CHAR(5) NOT NULL");
+      f.checkString("greatest('show', 'on', 'earth')", "show ",
+          "CHAR(5) NOT NULL");
+      f.checkScalar("greatest(12, CAST(NULL AS INTEGER), 3)", "12",
+          "INTEGER");
+      f.checkScalar("greatest(false, true)", true, "BOOLEAN NOT NULL");
+
+      final SqlOperatorFixture f12 = f.forOracle(SqlConformanceEnum.ORACLE_12);
+      f12.checkString("greatest('on', 'earth')", "on", "VARCHAR(5) NOT NULL");
+      f12.checkString("greatest('show', 'on', 'earth')", "show",
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10770]: You should add negative test if the function returns null or fails

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 6]
```diff
@@ -10772,8 +10794,30 @@ void assertSubFunReturns(boolean binary, String s, int start,
...
+      f.checkString("least('on', 'earth')", "earth", "CHAR(5) NOT NULL");
+      f.checkString("least('show', 'on', 'earth')", "earth",
+          "CHAR(5) NOT NULL");
+      f.checkScalar("least(12, CAST(NULL AS INTEGER), 3)", "3",
+          "INTEGER");
+      f.checkScalar("least(false, true)", false, "BOOLEAN NOT NULL");
+
+      final SqlOperatorFixture f12 = f.forOracle(SqlConformanceEnum.ORACLE_12);
+      f12.checkString("least('on', 'earth')", "earth", "VARCHAR(5) NOT NULL");
+      f12.checkString("least('show', 'on', 'earth')", "earth",
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10817]: above all

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 7]
```diff
@@ -10747,8 +10747,30 @@ void assertSubFunReturns(boolean binary, String s, int start,
           "VARCHAR(5) NOT NULL");
     };
     final List<SqlLibrary> libraries =
-        list(SqlLibrary.BIG_QUERY, SqlLibrary.ORACLE, SqlLibrary.REDSHIFT,
-            SqlLibrary.SPARK);
+        list(SqlLibrary.BIG_QUERY, SqlLibrary.ORACLE, SqlLibrary.SPARK);
+    f0.forEachLibrary(libraries, consumer);
+  }
+
+  @Test void testGreatestPgFunc() {
```
Comment (caicancai) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10754]: Add jira link

File: core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java (hunk 4 above)
Comment (normanj-bitquill) [reply core/src/main/java/org/apache/calcite/sql2rel/StandardConvertletTable.java:1907]: You are right. Let expressions would avoid repeated evaluation of the arguments. Does Calcite support LET expressions?

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 6 above)
Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: Added a test where both arguments are null.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 7 above)
Comment (normanj-bitquill) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10756]: Added Javadoc comments for the Jira ticket.

**Review by caicancai — APPROVED** [review]
LGTM

**Comment by normanj-bitquill** [comment]
The commits have been squashed.

**Review by mihaibudiu — APPROVED** [review]
